src/tools/sign-oci-layer-root-hashes/src/sign.rs

```diff
@@ -36,6 +50,137 @@ struct LayerInfo {
+                annotations.insert(IMAGE_LAYER_ROOT_HASH_LABEL.to_string(), layer.root_hash.clone());
+                annotations.insert(IMAGE_LAYER_SIGNATURE_LABEL.to_string(), layer.signature.clone());
+                annotations.insert(SIGNATURE_FILE_NAME.to_string(), file_name.clone());
+
+                // Push each signature blob to the repository
+                client.push_blob(
+                    image_ref,
+                    &blob_bytes,
+                    &blob_digest,
+                ).await
+                .context("Blob pushed failed")?;
+
+                // Create the signature descriptor
+                Ok::<OciDescriptor, Error>(OciDescriptor {
+                    media_type: SIGNATURE_MEDIA_TYPE.to_string(),
+                    digest: blob_digest,
+                    size: blob_bytes.len() as i64,
+                    annotations: Some(annotations),
+                    urls: None,
+                    ..Default::default()
+                })
+            }
+        })
+    ).await?
+    .into_iter()
+    .collect();
+    // 3. Create the signature manifest with the subject descriptor, empty config, artifact type, and annotations
+    let mut annotations = BTreeMap::new();
+    annotations.insert(IMAGE_NAME_LABEL.to_string(), image.name.clone());
+    annotations.insert(
+        "org.opencontainers.image.created".to_string(),
+        chrono::Utc::now().to_rfc3339(),
+    );
+    // Empty Descriptor for config, as per OCI spec, this is required for the manifest, refer to https://github.com/opencontainers/image-spec/blob/main/manifest.md#guidance-for-an-empty-descriptor
+    let empty_config_digest = "sha256:44136fa355b3678a1146ad16f7e8649e94fb4fc21fe77e8310c060f61caaff8a";
+    let config_descriptor = OciDescriptor {
+        media_type: "application/vnd.oci.empty.v1+json".to_string(),
+        digest:empty_config_digest.to_string(),
+        size: 2,
+        annotations: None,
+        urls: None,
+    };
+    // Ensure the empty config blob exists in the registry
+    client.push_blob(
+        &image_ref,
+        b"{}",
+        empty_config_digest,
+    ).await.context("Failed to push empty config blob")?;
+    let sig_manifest = oci_client::manifest::OciImageManifest {
+        schema_version: 2,
+        media_type: Some(oci_client::manifest::OCI_IMAGE_MEDIA_TYPE.to_string()),
+        artifact_type: Some(SIGNATURE_ARTIFACT_TYPE.to_string()),
+        config: config_descriptor,
+        layers: sig_descriptors.clone(),
+        subject: Some(subject_descriptor.clone()),
+        annotations: Some(annotations.clone()),
+    };
+
+    // 4. Push the signature manifest to a new reference (e.g., with a ".erofs.sig" suffix)
+    let sig_ref: Reference = Reference::with_tag(
+            image_ref.registry().to_string(),
+            image_ref.repository().to_string(),
+            format!("{}.erofs.sig", image_ref.tag().unwrap()),
```

Comment:
Wonder if we should have something more variable, so multiple signatures can be attached?

Reply:
Append the serial number of the signer certificate to the tag to ensure uniqueness and can be used by snapshotter to pick signatures for a specific signer.